Implements re-buildings and dependencies building.

jtkech opened the pull request:
Right now you need to manually compile `Orchard.ContentTypes.Abstractions` because it's not a module which can be compiled before (because of the modules loading order) and it's not compiled through the web app. Otherwise seems to works but needs to be refactored ...

Note: For infos, during my tests if i wanted to load the `Orchard.Contents` before, i needed to specify in the module.txt a dependency on `Orchard.Contents`. If i only specify `Contents`, as seen with some modules, it seems that it doesn't work.

Best

jtkech commented:
@sebastienros 

Very very experimental but now dependencies are analyzed and compiled if needed. Need to be refactored and can be improved. E.g there is the double extensions loading and right now all dependencies are analyzed even if it's a library which is already part of the orchard.web project ...

- So, the double extensions loading needs to be fixed.

- And i'll detect which libraries are already part of the orchard.web project, i already do this in `ExtensionLibraryService`.

Note: for the `Orchard.Contents` dependency which doesn't work if you use the extension name `Contents`, it's because this name is not parsed in the manifest file. In `ExtensionHarvester.ParseManifest()` line 178, i can see that `field[0].ToLowerInvariant() = "name"` and `NameSection = "name"`, but the switch / case doesn't work, `manifest.Add(NameSection, field[1])` is not executed. This for the `Orchard.Contents` module, seems not to be the case for others, very weird!

Best

Jetski5822 commented on src/Orchard.Environment.Extensions/Compilers/CSharpExtensionCompiler.cs:
```diff
@@ -272,5 +320,49 @@ private static Command RunCsc(string[] cscArgs)
+            }
+
+            // find inputs that are newer than the earliest output
+            var newInputs = inputs.Where(p => File.GetLastWriteTimeUtc(p) >= minDateUtc);
```
Instead of where, just use .Any()

jtkech commented:
@Jetski5822 done, thanks.
Need to sleep now :)

jtkech commented:
@sebastienros 

I've done some updates, e.g fixed the 2nd extensions loading by using `ApplicationPartManager` in place of `IExtensionManager`. For the last one i couldn't inject it directly through the constructor because of a circular dependency. I think it's because `IExtensionManager` injects extensions loaders, one loader injects `IExtensionLibraryService` which can't inject `IExtensionManager`.

 just a note on the last commit above because it's a so weird issue on string parsing! In the `Module.txt` the name section was not correctly parsed for 2 modules `Orchard.Contents` and `Orchard.Setup`. This explains the issue with a dependency on `Orchard.Contents` by using its name.

I could fix it by re-creating the 2 manifest files. Maybe because of some non visible special characters. That said, in `ExtensionHarvester.ParseManifest()` here the tests i've done before on the `field[0]` variable:

    // for Orchard.Contents & Orchard.Setup, on the name section line
    // while debugging, i can see that field[0] = "Name", then

    field[0].StartsWith("Name") // return true
    field[0].EndsWith("Name") // return true
    field[0].Contains("Name") // return true
    field[0].Equals("Name") // but this one return false!

Very weird but that's why the following `switch / case` was not working.

Best

jtkech commented:
@sebastienros 

In the last update:

- `csc.dll` is embedded in orchard.web and copied on building as specified in project.json. So, right now, maybe with vs tools you still have to copy it manually, but it should work in the future.

- `csc.runtimeconfig.json` is automatically generated at runtime, you no longer need to copy it.

Notes & todo:

- There are still 2 conditions which don't trigger compilation on startup, if a file has been removed or added but with an old date because of a copy / paste (dotnet build use an incremental cache file).

- See if it's worth to implement the above, i could re-use a response file which memorize all compilation IO between 2 startups, i've removed the generation of this file to minimize IO operations.

- To adapt the code, see in parallel how Orchard 1 do things. Seems that some implementations can be done separately e.g files watching ... So, right now, i still focus on compilation and on startup.

- After testing: refactoring, helpers / extensions ...

Best

jtkech commented:
@sebastienros 

Just done the checking on added / deleted compile inputs since the last compilation (because we can't only rely on dates). I've re-integrated the compilation response file that i use as a cache.

There are many other things to do, file watching, populate probing folders, more loading coordination (e.g maybe with artifacts folders) ... But, only related to compilation, i think we have about what we need.

But, right now, only on a dev environment, because i don't know yet how it will behave e.g in a published context, how all paths will be resolved ...

I think it's good that you test it on your environment before there are too many things.

First i need to rebase and resolve conflicts, i'll do it asap.

Best.

jtkech commented:
See my previous comment and its update section

jtkech commented:
Added diagnostics messages (string list) returned from the csc compiler. Right now i only init a local variable for testing but we will be able to use it for logging.

    var compiler = new CSharpExtensionCompiler();
    var success = compiler.Compile(projectContext, "Debug");
    var diagnostics = compiler.Diagnostics;

- If success and no diagnostics => all is ok, info compilation succeeded.
- If success and some diagnostics => compilation ok but some warnings.
- If not succees => compilation failure with some errors in diagnostics.

GVX111 commented:
on conpilation making error 
`{"The filename or extension is too long"}`

jtkech commented:
Thanks for the infos.

But i can't repro, just re-tried by deleting all modules (which now include core modules) and themes bin folders before starting the app, and i had no issue.

Need to be followed, if you can grab more infos ...

Best.

jtkech commented:
@GVX111 

So, i can't repro. The only change related to file path i've made is on the working directory of the csc command. This to specify the current project directory in place of always using the running web app directory. But seems to be useless on my side and maybe not so good. So, i just reverted this little change. Can you try it with this last update?

Best.

GVX111 commented:
var result = RunCsc(allArgs.ToArray())
                .WorkingDirectory(Directory.GetCurrentDirectory())
                .OnErrorLine(line => OnOutputLine(line, context, Diagnostics))
                .OnOutputLine(line => OnOutputLine(line, context, Diagnostics))
                .Execute();
i have this but same error. :|

jtkech commented:
Sorry, i can't repro.

So, it seems that in your case there is a file with a too long path. Where do you see the error message? While debugging, in the new provided Diagnostics message list?

Try to put some breakpoints, e.g when running the csc command to see the allArgs collection. Much simpler, in each modules `obj` folder, a `dotnet-compile-csc.rsp` file is generated. See in this file if a too long path is present.

Best

GVX111 commented:
[text.txt](https://github.com/OrchardCMS/Orchard2/files/290141/text.txt)
this is in .rsp

jtkech commented:
Ok, i think i understand your scenario.

Dynamic compilation is only intended to compile extensions. Right now extensions are projects which are harvested from `Orchard.Web/Modules` and `Orchard.Web/Themes` (others folders could be added but by code / config files).

Then a module / extension project can depend on another module / extension. It can also depend on another project library which is not an extension, but in this case this library need to be an ambient library. This means that this library need to be in the Orchard.Web project dependencies.

This is because ambient libraries are loaded naturally and extensions libraries are loaded dynamically. So, if a project is not an ambient library neither an extension, it's assemblies will not be loaded. So, a module / extension can't depend on it.

It seems this is your scenario because in your file i can see that we try to compile the `src\Orchard.Site` project which is not an ambient library (because not precompiled with the app) and which is not an extension (because not in the Modules / Themes folder).

Best.

jtkech commented:
Anyway i got the same issue when trying as planned to run in a published env. After dotnet publish (with an adapted project.json) all is in a `publish` subfolder of `orchard.web/bin`. So, i have long paths as:

`"\"c:\\jtkech\\Orchard2\\src\\Orchard.Web\\bin\\Debug\\netstandardapp1.5\\publish\\Modules\\Orchard.ContentTypes.Abstractions\\bin\\Debug\\netstandard1.5\\Orchard.ContentTypes.Abstractions.dll\""`

But i fixed it by passing to the csc compiler the `dotnet-compile-csc.rsp` file (as dotnet do) this in place of passing arguments as an array of string (seems to be limited this way).

@GVX111 thanks to have pointed out this issue, i'll update my PR soon.

Best.

jtkech commented:
@sebastienros 

So, i've got something working under what i assume is a published context. I needed to deal with ambient orchard libraries which are not resolved by the project model api when project files are not there.

Because i still work on my local machine, nugets are resolved as before. Maybe, in a prod env, they will be naturally resolved from the web app folder, need to be investigated ...


**Maybe good to give it a try** as it is in your environment before there will be too many things.

Best.

jtkech commented:
@sebastienros here my last update to work in a published context.

- When in a `dev environment`, when a package is used by a module but not by the web app, it is stored in the `module/lib` folder. Currently, this only happens for the `Orchard.Logging.Console` which uses `Microsoft.Extensions.Logging.Console.dll`.

- When publishing the `publishOptions` specify to copy, among others, the `Modules` folder (include `bin` & `lib` folders, exclude `obj` folder ...), see `project.json`.

- In a `"published" context`, when the path of a package is not resolved, we try to search in the web app `root` directory, then in the `./refs` folder, then we fallback to the `module/lib` folder.

- In a `"published" context`, when the path of an ambient orchard library is not resolved, we fallback to the web app `root` directory.

Best

jtkech commented:
@sebastienros 

- 1st, without my PR, with the last master branch for some view i needed to comment out in 3 places the new `//[Admin]` attribute (not included in this PR).

- I've seen that now an orchard library can be neither ambient nor an extension but used as a dependency by an extension. So i needed to also compile / load them, and also store them in the module probing folders (with other module specific packages). This to work in a published context.

I also seen that core modules are with others. Wow, now so many things to dynamically compile / load.

Best.

sebastienros commented:
Testing now ... for the [Admin] issue it must be that you need to clean your repos and do a dotnet restore. The attribute is in a different assembly now.

sebastienros commented:
1- Probing folders should be .gitignore'd
2- Shouldn't there be a single probing folder, like in O1?

jtkech commented:
@sebastienros 

1- You're right.

2 - You're certainly right. Is it possible that 2 modules use the same package dll but not the same version?

There are so many things, i was also thinking about artifact loader, but here maybe better to change VStudio xproj files ...


Right now i want to see how it behaves for you and if it's worth to continue.

Best

jtkech commented:
@sebastienros 

Regardless my PR, because the `AdminAttribute` (of type `AdminFilter`) inherit from `ServiceFilterAttribute`, as i understand `AdminFilter` need to be registered as a service.

Not sure it's a good fix, but to be able to go to the `/orchard.themes/admin/index` view, i needed to add the following to get it working, then i was at least redirected to the login view.

So, i've created this file `Orchard.Admin/Module.cs` (not sure it's the right place)

    using Microsoft.Extensions.DependencyInjection;
    using Microsoft.Extensions.DependencyInjection.Extensions;
    using Orchard.DependencyInjection;

    namespace Orchard.Admin
    {
        public class Module : IModule
        {
            public Module()
            {
            }

            public void Configure(IServiceCollection serviceCollection)
            {
                serviceCollection.TryAddScoped<AdminFilter>();
            }
        }
    }

Best.

jtkech commented:
@sebastienros 

- Use a single probing folder `App_Data/Dependencies`. Currently i store all extensions (this is not necessary) and all their dependencies which can be packages / orchard libraries / extensions.

- Select the `Debug / Release` configuration based on the compilation options of the running assembly.

- TODO: If no source files (or just test if no project json files), then don't compile, still try to load the extension, manage bin / probing folders ...

Best

jtkech commented:
@sebastienros 

- So, we know if an extension is ambient (maybe now avoided) then we do an `ambient loading`.

- We know if project files are there and we compile if needed and we do a `dynamic loading`.

- If no project files, we would need to do a `precompiled loading`. But here we still need to know the dependencies to load e.g from the probing folder. Here there are different strategies (see below).

**TODO**:

- In place of trying to load with each loader, i think i need to do a `loader coordinator`.

- For `precompiled loading`, define a way to know dependencies to load. One solution is to always have `project.json and project.lock.json` files in extension folders.

- Another solution is to compile extensions with the `preserve context` option (the assembly will embed `Project.lock.json` infos) (we can force this option). Then we could use `dependencyContext`.

- A last solution is to manage dependencies files but maybe better to let dotnet api do it for us.

I would opt to compile extensions with the `preserve context` option, then we don't rely on `project json files` for `precompiled loading`. But this produces bigger dlls, e.g `300 kb` in place of `30 kb`.

Best

jtkech commented:
- So, i re-use different loaders which use related loading methods through the `ExtensionLibraryService`.

- `Ambient` and `dynamic` loading done, need to implement `precompiled` loading.

- Need refactoring to share code between loading methods.

sebastienros commented:
Have you tried within Visual Studio ? Like changing a module, compiling from Vs ,then starting the site to ensure the change is taken into account?

jtkech commented:
I've tried all of this but only through vscode. I'll try it under vstudio.

Right now i write precompiled loading if there are no project files. E.g i have to resolve the module compilation output path without project files. This to see the dll in the module bin folder ...

jtkech commented:
@sebastienros 

- By doing this in O2, i begin to better understand what is done in O1.

- With `precompiled loading` (here means no project files, so, anyway we can't compile), i see new scenarios, e.g when a dependent project needs to be compiled. Just seen it's not implemented in O1:

    `Logger.Information("Extension \"{0}\" will not be loaded as pre-compiled extension because
        one or more referenced extension is dynamically compiled", extension.Id);`

- We could do this and it's naturally done because, normally, the dependent project is compiled before. But here maybe a problem of version ...

- **If 2 modules use the same package but not the same version, can we load both versions**? If so, we still need module probing folders. Maybe i would need to check versions (not only names & dates).

- If 2 versions are not allowed, a shared probing folder makes more sense. But in O1, when you delete the `dependencies folder`, it is re-created, even when published. This can be done because dlls also are in modules bin folders.

- Then, not sure we need the shared probing folder we would need to maintain. In O1, seems to be used as a global "dependency context" for all extensions ... Maybe useful if we can say when published don't delete it but not sure.

- Hope we don't still rely on some dotnet sdk assemblies, but if so i think we will be able to embed them.


Best.

jtkech commented:
@sebastienros just for infos

- **Ambient loading**: Ok.

- **Dynamic loading**: Ok for me, i've re-done some tests by updating projects / dependencies, deleting  outputs in binary / probing folders ... So seems ok for a dev environment.

- **Precompiled loading**: I was assuming that a module is precompiled when there are no project files. Then, if precompiled with the `preserveContext` option, we can use `dependencyContext` on it.

- **Drawback**: When you dyn compile a module with the `preserveContext` option, you can't have a dependency on a precompiled module which have no project files.

I think i'll revert back to lighter dlls without using this option. One solution i see is to always have project files. A precompiled module would be the one that doesn't have any source files (*.cs), only its output binaries and still its 2 project json files. Seems to work in my 1st tests on this.

Best.

sebastienros commented:
I am wondering if we should have a requirement on what the output path should be. I had to change them all to point to the correct artifacts folder as they are in a subfolder and were copy-pasted. I will ask the VS Tools team.

jtkech commented:
@GVX111

Thanks for the feedback. I know vstudio tools use xproj but not tried it yet. But when we dyn compile we use the project model which uses project json files to e.g resolve output paths ..., so i was assuming it would be ok. Did you try with the last update?

Oh maybe because in your context, at runtime, some resolved paths point to artifacts that we can't access. **So, next step, see what happens under the vstudio context**, i'll try it asap.

@sebastienros 

That said i began to add debug console message, cool to know when a dynamic compilation succeeded, was skipped because already compiled ... I'll post a screenhost.

Best

GVX111 commented:
@jtkech u did awesome job. Yes i checked lastest and everything are ok but i think it need optimization because first compilation of dyn exts need much more time than i expected.

sebastienros commented:
Crazy suggestion, do parallel compilation based on project dependencies ;)
@GVX111 please compare the time it takes on the command line, because even there it's very slow. Like 2-3 seconds per project.

jtkech commented:
@sebastienros 

Just tried with vstudio tools, indeed @GVX111 was right about artifacts bin folders. Maybe i have some solutions (see my last comments on gitter), will try it asap.

Best

jtkech commented:
- Just added a stopwatch to see dyn compilation time.

- @GVX111 thanks for the suggestion but didn't tried yet to do any optimization. For compile time, maybe we are faster because we start only one process for csc.dll. I think we are slower to see if we need to compile, dotnet use some caching but at the workspace level, need to see again this area.

- That said we have to know that, between 2 dyn compilations, we update probing folders, and, especially, we load the extension dependencies assemblies.

- @sebastienros just seen your suggestion ;). Maybe doable, so, we have to check not to compile in parallel the same project, or 2 projects that have a shared dependency that need to be compiled and / or copied to a shared probing folder ...

Best.

GVX111 commented:
@sebastienros @jtkech I thought about parallel but it will be difficult because maybe some Extensions will be dependency. Need Ordering extension. I don't know if @jtkech did already if extension referenced to another extension, referenced extension doesn't need again compilation. If it is already done need only order by dependency.

GVX111 commented:
@sebastienros  for me from command line compilation need 3-4 seconds per project. it is very slow. I have ssd and HQ processor I don't think it is laptop requirement problem. It will be much more faster with parallel but first of all need ordering extensions.

jtkech commented:
@GVX111 

We already recursively compile extensions, so, if an extension depends on another project it is compiled before. Note: In the compiler, just changed the way a precompiled project is detected. Now it's when there are no source files but still with project files (needed by the project model api).

Anyway in `ApplicationBuilderExtensions` the `applicationPartManager` is populated with `AvailableFeatures()` which, in `ExtensionManager`, uses `.OrderByDependenciesAndPriorities`. So, extension are already loaded / compiled in an order based on dependencies.

So, for parallel compilation, maybe we only need a specific locking system to prevent parallel compilation of the same project / dependency and / or storing the same assembly in the shared probing folder.

Best.

jtkech commented:
@sebastienros 

- Parallel compilation done, seems to work (can be easily reverted).

- We already recursively compile according to dependencies. So just added a locking system based on the project name which is currently compiled.

- A lock is alo used in 2 other places where we could potentially write the same file at the same time.

- Under VS, to see compilation messages in the debug console, i use the `IIS Express` start command.


Best.

Jetski5822 commented:
@jtkech did you see any perf difference with Parallel Compilation?

Jetski5822 commented on src/Orchard.Environment.Extensions/Compilers/CSharpExtensionCompiler.cs:
```diff
@@ -137,19 +299,146 @@ public bool Compile(ProjectContext context, string config, string buildBasePath)
+            }
+            else
+            {
+                Debug.WriteLine(String.Format($"{context.RootProject.Identity.Name}: Dynamic compilation failed."));
```
no need for string.format here

jtkech replied:
Indeed, done.

Jetski5822 commented on src/Orchard.Environment.Extensions/Compilers/CSharpExtensionCompiler.cs:
```diff
@@ -137,19 +299,146 @@ public bool Compile(ProjectContext context, string config, string buildBasePath)
+            }
+            else if (result.ExitCode == 0 && Diagnostics.Any())
+            {
+                Debug.WriteLine(String.Format($"{context.RootProject.Identity.Name}: Dynamic compilation succeeded but has warnings."));
```
no need for string.format here

jtkech replied:
Done.

Jetski5822 commented on src/Orchard.Environment.Extensions/Compilers/CSharpExtensionCompiler.cs:
```diff
@@ -137,19 +299,146 @@ public bool Compile(ProjectContext context, string config, string buildBasePath)
+            {
+                Debug.WriteLine(String.Format($"{context.RootProject.Identity.Name}: Dynamic compilation succeeded."));
+                Debug.WriteLine($"0 Warning(s)");
+                Debug.WriteLine($"0 Error(s)");
```
remove unneeded string interpolation

jtkech replied:
Oh yes but since i use parallel compilation, here i needed to specify again the project name with the compilation result message.

Jetski5822 commented on src/Orchard.Environment.Extensions/Compilers/CSharpExtensionCompiler.cs:
```diff
@@ -137,19 +299,146 @@ public bool Compile(ProjectContext context, string config, string buildBasePath)
+
+            if (result.ExitCode == 0 && !Diagnostics.Any())
+            {
+                Debug.WriteLine(String.Format($"{context.RootProject.Identity.Name}: Dynamic compilation succeeded."));
```
no need for string.format here

jtkech replied:
Done.

Jetski5822 commented on src/Orchard.Environment.Extensions/Compilers/CSharpExtensionCompiler.cs:
```diff
@@ -137,19 +299,146 @@ public bool Compile(ProjectContext context, string config, string buildBasePath)
-            return result.ExitCode == 0;
+            Debug.WriteLine(String.Empty);
+
+            if (result.ExitCode == 0 && !Diagnostics.Any())
```
if Diagnostics is a list, use Diagnostics.Length > 0 - its faster.

jtkech replied:
Ok, but here there is no Length property, so i've used the Count property. Done.

Jetski5822 commented on src/Orchard.Environment.Extensions/Compilers/CSharpExtensionCompiler.cs:
```diff
@@ -137,19 +299,146 @@ public bool Compile(ProjectContext context, string config, string buildBasePath)
+                }
+            }
+
+            Debug.WriteLine(String.Format($"{context.RootProject.Identity.Name}: Dynamic compiling for {context.TargetFramework.DotNetFrameworkName}"));
```
remove string.format

jtkech replied:
Done.

Jetski5822 commented on src/Orchard.Environment.Extensions/Compilers/CSharpExtensionCompiler.cs:
```diff
@@ -137,19 +299,146 @@ public bool Compile(ProjectContext context, string config, string buildBasePath)
+
+                    if (!prevInputs.Except(newInputs).Any() && ! newInputs.Except(prevInputs).Any())
+                    {
+                        Debug.WriteLine(String.Format($"{context.RootProject.Identity.Name}: Previously compiled, skipping dynamic compilation."));
```
remove string.format

jtkech replied:
Done.

Jetski5822 commented on src/Orchard.Environment.Extensions/Compilers/CSharpExtensionCompiler.cs:
```diff
@@ -137,19 +299,146 @@ public bool Compile(ProjectContext context, string config, string buildBasePath)
+
+            inputs.AddRange(sourceFiles);
+            inputs.AddRange(references);
+            outputs.AddRange(outputPaths.CompilationFiles.All());
```
var outputs = outputPaths.CompilationFiles.All() on line 305?

jtkech replied:
Yes, All() is a public method of the Microsoft.DotNet.ProjectModel.CompilationOutputFiles class, then you get an IEnumerable<string> where you have the .dll output path and also e.g the .pdb output path.

Jetski5822 commented on src/Orchard.Environment.Extensions/Compilers/CSharpExtensionCompiler.cs:
```diff
@@ -75,46 +138,147 @@ public bool Compile(ProjectContext context, string config, string buildBasePath)
             {
-                var allExports = exporter.GetAllExports().ToList();
-                var dependencyContext = new DependencyContextBuilder().Build(compilationOptions,
+                var allExports = exporter.GetAllExports().Where(x => x.Library.Compatible).ToList();
```
Maybe extension method to say, GetAllCompatibleExports()?

jtkech replied:
Yes, you're right, i'll do it. Maybe other places to do the same ... I planed to do this kind of things but because i'm not sure all of this will be used ...

Jetski5822 commented on src/Orchard.Environment.Extensions/Compilers/CSharpExtensionCompiler.cs:
```diff
@@ -75,46 +138,147 @@ public bool Compile(ProjectContext context, string config, string buildBasePath)
+                // Check for an unresolved package
+                else if (package != null && !package.Resolved)
+                {
+                    foreach (var assembly in package.CompileTimeAssemblies)
```
Could this be parallelized? making 'references' as a ConcurrentDictionary

jtkech replied:
Oh yes, why not, will try it.

Jetski5822 commented on src/Orchard.Environment.Extensions/Compilers/CSharpExtensionCompiler.cs:
```diff
@@ -1,65 +1,128 @@
+            if (_ambientLibraries.IsEmpty)
+            {
+                var libraries = DependencyContext.Default.CompileLibraries
+                    .Where(x => x.Type == LibraryType.Project.ToString().ToLowerInvariant());
```
string.Equals(x.Type, LibraryType.Project.ToString(), StringComparision.Original  <- You know wha I mean (I forgot the last bit))

jtkech replied:
Done.

Jetski5822 commented on src/Orchard.Environment.Extensions/Compilers/CSharpExtensionCompiler.cs:
```diff
@@ -75,46 +138,147 @@ public bool Compile(ProjectContext context, string config, string buildBasePath)
+            // Check again if already compiled, here through the dependency graph
+            if (_compiledLibraries.TryGetValue(context.RootProject.Identity.Name, out compilationResult))
+            {
+                return _compiledLibraries[context.RootProject.Identity.Name];
```
return compilationResult;  (You have the result from line 248)

jtkech replied:
Indeed, done.

Jetski5822 commented on src/Orchard.Environment.Extensions/Compilers/CSharpExtensionCompiler.cs:
```diff
@@ -1,65 +1,128 @@
             {
                 // We got an unresolved dependency or missing framework. Don't continue the compilation.
-                return false;
+                return _compiledLibraries[context.RootProject.Identity.Name] = false;
```
Does this compile?

jtkech replied:
Here, we skip the extension / dependency project compilation. Then, right now, it depends on an already compiled assembly in a binary / probing folder ...

Jetski5822 commented on src/Orchard.Environment.Extensions/Compilers/CSharpExtensionCompiler.cs:
```diff
@@ -75,46 +138,147 @@ public bool Compile(ProjectContext context, string config, string buildBasePath)
             if (String.IsNullOrEmpty(intermediateOutputPath))
             {
-                return false;
+                return _compiledLibraries[context.RootProject.Identity.Name] = false;
```
Does this compile?

jtkech replied:
Idem.

Jetski5822 commented on src/Orchard.Environment.Extensions/Compilers/CSharpExtensionCompiler.cs:
```diff
@@ -267,10 +555,38 @@ private static string GetLanguageVersion(string languageVersion)
+                {
+                    minDateUtc = lastWriteTimeUtc;
+                }
+            }
```
replace this with....

var minDateUtc = outputs.Min(output => File.GetLastWriteTimeUtc(output));

maybe?

jtkech replied:
Thanks, i like it, done.

Jetski5822 commented:
Awesome work so far man, left some notes, not sure if any of them are of use?

jtkech commented:
@Jetski5822 

Thanks a lot for all your insights, no time right now, i'll update my PR according to them this night.

> did you see any perf difference with Parallel Compilation?

No comparison, it's about 3 or 4 times faster.

Best.

jtkech commented:
@Jetski5822 

Just done your suggested extension and i also use an existing dotnet extension i didn't see: `ProjectName()`. I think we will be able to find / create other ones, but maybe better to wait ...

So, in the existing folder `Orchard.Environment.Extensions/Extensions/` i've created a new file `ProjectModelExtensions.cs`, this to put extensions related to the dotnet project model classes: `LibraryExporter`, `ProjectContext` ... But not sure it's the best place / name.

Best

jtkech commented:
@Jetski5822 

> Could this be parallelized? making 'references' as a ConcurrentDictionary

Just thinking about this suggestion. So, i think you mean to parallelize what we do in the `foreach (var dependency in dependencies)` loop (currently line 145). Notice that in this loop, when a dependency is a resolved library project, this is where we recursively call `Compile()` (currently line 162).

So, wow, each compilation, already called in parallel, could potentially call other compilations in parallel! Not sure how it could behave with the current locking system, but for sure i want to try it.

Best.

jtkech commented:
**Notes**

- After my last dotnet restore, many packages support netstandard1.6 but not 1.5 (or lower). So they are not compatible with netcoreapp1.0 which doesn't support netstandard1.6 ...

- Right now, for most of the dependencies in project.json files i target `1.0.0-rc2-final`. **This is not integrated in this PR**, it was just to keep working. I'll see how i can get it working with latest bits ...

**Last updates**

- Maybe due to the last packages updates but i've seen the case where we try to load an ambient assembly. I could fix it by using another collection to have a more complete list of ambient libraries.

- Began to re-integrate some stuff around resource files generation. At this time, when resources are defined in a given module, resource files are generated and outputed in this module bin folder. Then not sure what i have to do with these files, will see ... If no resource defined, behave as before.

Best.

jtkech commented:
@Jetski5822 

I got a solution that works with Core 1.0. Because i've done all the work on my branch, right now i put it here so you can review it. I don't understand how you could compile with core app 1.0, netstandard 1.6 and latest packages references because of some interfaces / classes / namespaces changes. Note that i've updated all my tools, dotnet core, sdk ... And i've used restore with no cache `dotnet restore --no-cache`.

- See https://docs.asp.net/en/latest/migration/rc2-to-rtm.html about the `AuthorizationHandler<TRequirement>` change.

- See https://github.com/aspnet/Mvc/blob/master/src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcRouteHandler.cs, we can't use anymore `new MvcRouteHandler()` without any arguments. Now there is only one `MvcRouteHandler()` per app that you can request as a service. Note: but i've seen that you can have many `MvcAttributeRouteHandler` but right now not sure what we can do with this.

- See https://github.com/aspnet/Mvc/blob/master/src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/ControllerModel.cs, the `ControllerModel` doesn't have anymore `RouteConstraints`, replaced with `RouteValues` which still means values that must be present.

- `HtmlString` is not in the same namespace ...

So, most of the changes was simple but not sure for few of them, e.g about `MvcRouteHandler`, maybe here we need a deeper analyze. At least it couldn't work as it was and now it works for me (not tested with multi tenants) and you can review it.

Best.

Jetski5822 commented on src/Orchard.Hosting.Web/Routing/ModuleAreaRouteConstraintApplicationModelProvider.cs:
```diff
@@ -31,7 +31,8 @@ public void OnProvidersExecuted(ApplicationModelProviderContext context)
                 {
-                    controller.RouteConstraints.Add(new AreaAttribute(feature.Descriptor.Id));
+                    var areaAttribute = new AreaAttribute(feature.Descriptor.Id);
+                    controller.RouteValues.Add(areaAttribute.RouteKey, areaAttribute.RouteValue);
```
Interesting, I did this....

controller.RouteValues.Add("area", feature.Descriptor.Id);

Jetski5822 commented on src/Orchard.Hosting.Web/Routing/OrchardRouterMiddleware.cs:
```diff
@@ -75,7 +75,7 @@ public RequestDelegate BuildTenantPipeline(ShellSettings shellSettings, IService
             var routeBuilder = new RouteBuilder(appBuilder)
             {
-                DefaultHandler = new MvcRouteHandler()
+                DefaultHandler = serviceProvider.GetRequiredService<MvcRouteHandler>()
```
Is this correct? Is the MvcRouteHandler not on the ApplicationServices?

Jetski5822 commented on src/Orchard.Security/AuthorizationHandlers/PermissionHandler.cs:
```diff
@@ -10,17 +12,18 @@ namespace Orchard.Security.AuthorizationHandlers
-                return;
             }
+
+            return Task.CompletedTask;
```
yup I did same

Jetski5822 commented on src/Orchard.Security/AuthorizationHandlers/PermissionHandler.cs:
```diff
@@ -10,17 +12,18 @@ namespace Orchard.Security.AuthorizationHandlers
     public class PermissionHandler : AuthorizationHandler<PermissionRequirement>
     {
-        protected override void Handle(AuthorizationContext context, PermissionRequirement requirement)
+        protected override Task HandleRequirementAsync(AuthorizationHandlerContext context, PermissionRequirement requirement)
```
I put async and used await on the below tasks

Jetski5822 commented:
Okay cool - So I would like to get this in to the main branch..... Whats left?

Jetski5822 commented:
Last thing.... Update 
$koreBuildZip="https://github.com/aspnet/KoreBuild/archive/1.0.0-rc2.zip"

to

$koreBuildZip="https://github.com/aspnet/KoreBuild/archive/1.0.0.zip"

alexbocharov commented:
@jtkech Good work!

@Jetski5822 
> Last thing.... Update 
>  $koreBuildZip="https://github.com/aspnet/KoreBuild/archive/1.0.0-rc2.zip"
>
> to
> 
> $koreBuildZip="https://github.com/aspnet/KoreBuild/archive/1.0.0.zip"

and dockerfile update to `preview2`

Jetski5822 commented:
@alexbocharov Yup, @jtkech , if you make those changes and are happy that whats in the PR is GTG, knowing we can make changes after.... then lets pull it.

alexbocharov commented:
@Jetski5822 I can fix a new PR

Jetski5822 commented:
@alexbocharov Okay, then we just need to know this is GTG and bam ill hit the button.... wake up @jtkech !!!

jtkech commented:
@Jetski5822 

As said it was 1st to keep it working and regarding some of your good remarks e.g about `MvcRouteHandler`, indeed we need more investigation ... No time right now but i'll take a look this night.

But as you said we will be able to do some updates afterwards in another PR.

For the `build.ps1` update, i remember i've done this at some point to update my tools. I could integrate it in my PR but can't do it immediately. But here also, maybe @alexbocharov can do it in another PR.

**So for me it's GTG** and it's as you want. For sure i plan some other updates for specific things but i will do other PRs. In fact, i think it's time to test this PR to see if it behaves as expected, there are already so many things on it and we can't yet see all possible scenarios ...

Best.

Jetski5822 commented:
Okay, merged...
